Add drag and drop functionality

ryantam626 opened the pull request:
As a user with many files to upload, it's probably quicker to just drag and drop from a file explorer (or something to that effect), however said functionality doesn't exist yet.

This PR:
- Implements the file drag and drop functionality
  - Abuse the `autodetect` lexer instead of parsing the file extension (or content) and picking the "correct" lexer which is
somewhat harder.
- Allow removal of 0th file
  - This is to allow file drag and drop user to delete the 0th file and proceed to submit the paste (otherwise 0th file being empty will block the submit paste action)

I also included some extra commit while I worked on this feature, let me know if you prefer me to split them out / dropped them all together :pray:

supakeen commented:
Wow, cool. 

A remark:
- If you could please fix the test cases; they call `make_application` so putting a default value on the new `debug` argument to it should fix them up.

I think you're planning to do another commit in this PR to deal with the 0th file issue? I see three options:

1. Do as you said in the commit.
2. Make it much more complicated by making the drop target the text input; replacing the one you drop on and adding additional ones if more are necessary.
3. Making `add_new_file` a bit smarter and having it automatically use up any 'empty' files.

I'm not favoring any of those in particular. And 2 and 3 can be done as followups outside this PR.

Small side-question: could we make the drop target clickable as well so I can select files with the file-picker or would that be a lot of work?

ryantam626 commented:
> If you could please fix the test cases; they call make_application so putting a default value on the new debug argument to it should fix them up.

Yes, my bad for forgetting to run tests :sweat_smile: 

> I think you're planning to do another commit in this PR to deal with the 0th file issue? I see three options:

I really like the 3rd option here, I can pick it up and drop the commit for changing the 0th file behaviour.

> Small side-question: could we make the drop target clickable as well so I can select files with the file-picker or would that be a lot of work?

Ah! This completely slipped my mind, let me investigate now, I don't think it would be super hard.

ryantam626 commented:
Done :raised_hands:

supakeen commented:
If you could add a changelog note and fix the lint please.

I'm currently checking out this PR locally :)

ryantam626 commented:
:facepalm: my bad for not linting locally.

On it now!

supakeen commented:
Cool, this works very well. However I don't think drag/drop is the most important input method so if we could move the container to below the normal paste-bar it should still be above the fold for most browsers but not push the main copy/paste area down :)

ryantam626 commented:
Would you mind having a look again please :pray: ? 
I rebased master out of habit, as such I took the opportunity to clean up git history and removed all the unnecessary commits.

I have also re-ordered section as suggested and added a small CSS for styling fix.

supakeen commented:
Love it <3. Merging.

supakeen commented:
I'll be deploying master to `bpa.st` later today; then do a release after the weekend in case any bugs are found.
